### ISSUE TITLE
Vega-functions: spec files

### ISSUE THREAD

**Issue by martinRenou** [vega/vega#1938]
cc. @maartenbreddels

We are working on a Python -> Vega-expression transpilation package: https://github.com/QuantStack/py2vega

We provide [mocking Python functions](https://github.com/QuantStack/py2vega/blob/master/py2vega/functions/string.py) for all vega functions. In order to easily keep them up to date with vega it would be super useful to have spec files (JSON files?) for those functions. This spec could be used for automatically generating our code, but it would also be useful for generating the [Vega documentation](https://github.com/vega/vega/blob/master/docs/docs/expressions.md).

It could look like the following:
```
{
    "string": [
        {"name": "indexof", "args": ["string", "substring"], "doc": "Returns the first index of substring in the input string."},
        [...]
    ],
    [...]
}
```

**Comment by domoritz** [comment]
Would you be able to work with a typescript declaration file?

**Comment by martinRenou** [comment]
It would not be really usable... Don't you think?

**Comment by domoritz** [comment]
It'd be easier to maintain, testable, and standardized. I suppose it's a bit trickier to parse, though.

**Comment by martinRenou** [comment]
Yes, it would be tricky to parse. On your side for the documentation and on our side for the code generation.

**Comment by domoritz** [comment]
One option would be to extend our schema generator (https://github.com/vega/ts-json-schema-generator) to support functions. Then we'd get the best of both worlds.

**Comment by domoritz** [comment]
Btw, your project sounds similar to what @saulshanabrook is working on. Is it related?

**Comment by martinRenou** [comment]
It might be similar work, what are you working on @saulshanabrook? It would be nice to join efforts if we have the same goal :)

I know the Altair project has done something similar, not really using the same approach though (See https://github.com/altair-viz/altair/blob/master/altair/expr/core.py).

**Comment by saulshanabrook** [comment]
@martinRenou we are doing something very similar! If you are around right now, we could chat:

> To join the video meeting, click this link: https://meet.google.com/iiu-rryx-stt
> Otherwise, to join by phone, dial +1 478-292-3994 and enter this PIN: 268 736 605#

We are [moving all our work](https://github.com/Quansight/ibis-vega-transform/pull/11) into https://github.com/Quansight/ibis-vega-transform  from https://github.com/quansight/jupyterlab-omnisci.

The high level idea is to push the transforms from Vega to the backend, so that we can support interactive exploration of large datasets backed by databases or pandas, using Altair.

**Comment by saulshanabrook** [comment]
Few notes from the discussion with @martinRenou and @maartenbreddels:

* They are doing the opposite  thing that we are. They want to translate Python expressions into vega code, by parsing the AST. We are doing the opposite, taking Vega expressions and compiling it to Python.
* We could maybe use both tools together for E2E testing, to make sure performing both transforms gave the same result in python. 


cc @ian-r-rose

**Comment by jheer** [comment]
I'm interpreting this issue as primarily a request for a standardized JSON specification file that documents all functions in the Vega expression language; that is, all functions exported by the vega-functions package. I'm also flagging this as "help-wanted", as this is a relatively straightforward (if tedious) task for a community contribution. Folks should join the conversation here if they are interested in working on this, as we will want to first come to agreement on the format of the specification file.

**Comment by hydrosquall** [comment]
I've seen doc generation libs (like TSDoc) for parsing JS files for exported functions. To do this they must extract the docstring + list of args and types for each function. 

It sounds like the data structure that powers doc generation could be used for this "expression list" JSON. Do we foresee uses for this list aside from Python code generation?

**Comment by dangotbanned** [comment]
@hydrosquall thought I'd mention a fairly recent addition to `altair` 

https://github.com/vega/altair/blob/82ec2692efab4ad8f47b39623f1769d90e62f6ae/tools/vega_expr.py

We're currently parsing the [markdown docs](https://vega.github.io/vega/docs/expressions/) to approximate signatures.
It has it's quirks, but the UX (especially statically) has improved quite a bit since https://github.com/vega/vega/issues/1938#issuecomment-523001195

**Comment by hydrosquall** [comment]
@dangotbanned oh, clever!!  Thanks for sharing 

I suppose if we have such a JSON we could next use it to generate the Markdown (after the intro text), otherwise the two will likely drift apart.
